File: modules/03-paralleization/tutorial.google-adk.ts
/**
 * Run:
 * cd modules && node --env-file=.env --experimental-strip-types ./paralleization/tutorial.google-adk.ts
 */
import {
  Gemini,
  InMemoryRunner,
  LlmAgent,
  ParallelAgent,
  SequentialAgent,
  isFinalResponse,
  stringifyContent,
  type Event,
} from "@google/adk";
import { createUserContent, type Content } from "@google/genai";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

export type GoogleAdkConfig = {
  apiKey: string;
  modelName: string;
  topic: string;
  appName: string;
  userId: string;
};

export type GoogleAdkResult = {
  report: string;
};

export function resolveGoogleAdkConfig(
  env: NodeJS.ProcessEnv = process.env,
): GoogleAdkConfig {
  const apiKey: string | undefined = env.GEMINI_API_KEY?.trim() || env.GOOGLE_API_KEY?.trim();
  if (!apiKey) {
    throw new Error("GEMINI_API_KEY or GOOGLE_API_KEY is missing.");
  }

  const config: GoogleAdkConfig = {
    apiKey,
    modelName: env.GOOGLE_MODEL?.trim() || "gemini-2.5-flash",
    topic: "Artificial Intelligence in Healthcare",
    appName: "parallelization-tutorial-adk",
    userId: "parallelization-tutorial-user",
  };
  return config;
}

export async function run(
  config: GoogleAdkConfig = resolveGoogleAdkConfig(process.env),
): Promise<GoogleAdkResult> {
  const model: Gemini = new Gemini({
    model: config.modelName,
    apiKey: config.apiKey,
  });

  const technologyAgent: LlmAgent = new LlmAgent({
    name: "TechResearcher",
    model,
    instruction: `Research the technology advances of ${config.topic}. Summarize in exactly one sentence.`,
  });

  const marketAgent: LlmAgent = new LlmAgent({
    name: "MarketResearcher",
    model,
    instruction: `Research the market impact of ${config.topic}. Summarize in exactly one sentence.`,
  });

  const parallelAgent: ParallelAgent = new ParallelAgent({
    name: "ParallelResearchers",
    subAgents: [technologyAgent, marketAgent],
  });

  const synthesisAgent: LlmAgent = new LlmAgent({
    name: "SynthesisAgent",
    model,
    instruction: "Combine both summaries into a concise markdown report.",
  });

  const workflow: SequentialAgent = new SequentialAgent({
    name: "GoogleAdkTutorialWorkflow",
    subAgents: [parallelAgent, synthesisAgent],
  });

  const runner: InMemoryRunner = new InMemoryRunner({
    agent: workflow,
    appName: config.appName,
  });

  const session: { id: string } = await runner.sessionService.createSession({
    appName: config.appName,
    userId: config.userId,
  });

  const prompt: string = `Create a short report for: ${config.topic}`;
  const userMessage: Content = createUserContent(prompt);
  const events: Array<Event> = [];

  for await (const event of runner.runAsync({
    userId: config.userId,
    sessionId: session.id,
    newMessage: userMessage,
  })) {
    events.push(event);
  }

  let report: string = "";
  for (let index: number = events.length - 1; index >= 0; index -= 1) {
    const event: Event = events[index]!;
    if (!isFinalResponse(event)) {
      continue;
    }

    const text: string = stringifyContent(event).trim();
    if (text.length > 0) {
      report = text;
      break;
    }
  }

  if (report.length === 0) {
    for (let index: number = events.length - 1; index >= 0; index -= 1) {
      const event: Event = events[index]!;
      const text: string = stringifyContent(event).trim();
      if (text.length > 0) {
        report = text;
        break;
      }
    }
  }

  if (report.length === 0) {
    throw new Error("Google ADK workflow completed without a text response.");
  }

  const result: GoogleAdkResult = { report };
  return result;
}

const isNodeTestContext: boolean = process.env.NODE_TEST_CONTEXT !== undefined;
const cliPath: string | undefined = process.argv[1];
if (!isNodeTestContext && cliPath) {
  const cliUrl: string = pathToFileURL(resolve(process.cwd(), cliPath)).href;
  if (cliUrl === import.meta.url) {
    void run()
      .then((result: GoogleAdkResult): void => {
        console.log(result.report);
        console.log(result);
      })
      .catch((error: unknown): void => {
        const message: string = error instanceof Error ? error.message : String(error);
        console.error(message);
        process.exitCode = 1;
      });
  }
}

File: 03-paralleization/tutorial_google_adk.py
import asyncio
import json
import os
from dataclasses import dataclass
from typing import Final

from google.adk.agents import LlmAgent, ParallelAgent, SequentialAgent
from google.adk.tools import google_search

from tutorial_types import StepLogger, TutorialStepResult

STEP_ID: Final[str] = "step-1-google-adk"
LIBRARY: Final[str] = "google-adk"


@dataclass(frozen=True)
class GoogleAdkConfig:
    api_key: str
    model_name: str
    topic: str


def resolve_google_adk_config() -> GoogleAdkConfig | None:
    gemini_api_key: str = os.getenv("GEMINI_API_KEY", "").strip()
    if gemini_api_key:
        return GoogleAdkConfig(
            api_key=gemini_api_key,
            model_name=os.getenv("GOOGLE_MODEL", "gemini-2.0-flash"),
            topic="Artificial Intelligence in Healthcare",
        )

    google_api_key: str = os.getenv("GOOGLE_API_KEY", "").strip()
    if google_api_key:
        return GoogleAdkConfig(
            api_key=google_api_key,
            model_name=os.getenv("GOOGLE_MODEL", "gemini-2.0-flash"),
            topic="Artificial Intelligence in Healthcare",
        )

    return None


async def run_google_adk_tutorial(log: StepLogger = print) -> TutorialStepResult:
    config: GoogleAdkConfig | None = resolve_google_adk_config()
    if config is None:
        return TutorialStepResult(
            step_id=STEP_ID,
            library=LIBRARY,
            status="skipped",
            message="GEMINI_API_KEY or GOOGLE_API_KEY is missing.",
        )

    os.environ["GOOGLE_API_KEY"] = config.api_key
    log("[google-adk] STEP 1/3 setup")

    tech_agent: LlmAgent = LlmAgent(
        name="TechResearcher",
        model=config.model_name,
        instruction=(
            f"Research the technological advancements of '{config.topic}'. "
            "Summarize in one sentence."
        ),
        tools=[google_search],
    )
    market_agent: LlmAgent = LlmAgent(
        name="MarketResearcher",
        model=config.model_name,
        instruction=(
            f"Research the market impact of '{config.topic}'. "
            "Summarize in one sentence."
        ),
        tools=[google_search],
    )
    parallel_agent: ParallelAgent = ParallelAgent(
        name="ParallelWebResearch",
        sub_agents=[tech_agent, market_agent],
    )
    synthesis_agent: LlmAgent = LlmAgent(
        name="SynthesisAgent",
        model=config.model_name,
        instruction="Combine the parallel research summaries into a concise report.",
    )
    workflow: SequentialAgent = SequentialAgent(
        name="ResearchSynthesisPipeline",
        sub_agents=[parallel_agent, synthesis_agent],
    )

    log("[google-adk] STEP 2/3 run")
    try:
        from google.adk.agents import InvocationContext

        context: InvocationContext = InvocationContext()
        events: list[str] = []
        async for event in workflow.run_async(context):
            events.append(str(event))

        output: str = "\n".join(events[-5:]) if events else "No ADK events emitted."
        return TutorialStepResult(
            step_id=STEP_ID,
            library=LIBRARY,
            status="completed",
            message="Google ADK tutorial completed.",
            output=output,
        )
    except Exception as exception:  # noqa: BLE001
        return TutorialStepResult(
            step_id=STEP_ID,
            library=LIBRARY,
            status="failed",
            message=str(exception),
        )


async def _main() -> None:
    result: TutorialStepResult = await run_google_adk_tutorial()
    print(json.dumps(result.__dict__, indent=2, ensure_ascii=True))


if __name__ == "__main__":
    asyncio.run(_main())
